component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java

```diff
@@ -0,0 +1,881 @@
+                    "Error when checking backup code enabled for the user : " + username, e);
+        } catch (AuthenticationFailedException e) {
+            throw new AuthenticationFailedException(
+                    "Authentication failed!. Cannot get the username from first step.", e);
+        } catch (URLBuilderException | URISyntaxException e) {
+            throw new AuthenticationFailedException("Error while building backup code page URL.", e);
+        }
+    }
+
+    @Override
+    protected void processAuthenticationResponse(HttpServletRequest request,
+                                                 HttpServletResponse response,
+                                                 AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        String token = request.getParameter(BACKUP_CODE);
+        AuthenticatedUser authenticatingUser =
+                (AuthenticatedUser) context.getProperty(BackupCodeAuthenticatorConstants.AUTHENTICATED_USER);
+
+        String username = authenticatingUser.toFullQualifiedUsername();
+        validateAccountLockStatusForLocalUser(context, username);
+
+        if (StringUtils.isBlank(token)) {
+            handleBackupCodeVerificationFail(context);
+            throw new AuthenticationFailedException(
+                    "Empty Backup code in the request. Authentication Failed for user: " +
+                            username);
+        }
+        try {
+            if (isInitialFederationAttempt(context)) {
+                if (!isValidTokenFederatedUser(token, context, username)) {
+                    throw new AuthenticationFailedException("Invalid Token. Authentication failed for federated user: "
+                            + username);
+                }
+            } else {
+
+                if (!isValidTokenLocalUser(token, username, context)) {
+                    handleBackupCodeVerificationFail(context);
+                    throw new AuthenticationFailedException(
+                            "Invalid Token. Authentication failed, user :  " + username);
+                }
+            }
+            if (StringUtils.isNotBlank(username)) {
+                AuthenticatedUser authenticatedUser = new AuthenticatedUser();
+                authenticatedUser.setAuthenticatedSubjectIdentifier(username);
+                authenticatedUser.setUserName(UserCoreUtil.removeDomainFromName(
+                        MultitenantUtils.getTenantAwareUsername(username)));
+                authenticatedUser.setUserStoreDomain(UserCoreUtil.extractDomainFromName(username));
+                authenticatedUser.setTenantDomain(MultitenantUtils.getTenantDomain(username));
+                context.setSubject(authenticatedUser);
+            } else {
+                context.setSubject(AuthenticatedUser.createLocalAuthenticatedUserFromSubjectIdentifier(username));
+            }
+        } catch (NumberFormatException e) {
```

Comment:
Do we need this? Do we need to treat this as a number? Just comparing as string is enough AFAIT.

Reply:
Yea fixed that